File: haskell/2023/Day4.hs
module Day4 where

import Data.Bifunctor
import Data.List
import Data.List.Split
import Data.Map.Strict (Map)
import Data.Map.Strict qualified as M
import Data.Maybe
import Data.Tuple.Common

parseCard :: String -> (Int, ([Int], [Int]))
parseCard = bimap (read . drop 5) (tmap (map read . filter (not . null) . splitOn " ") . toTuple . splitOn " | ") . toTuple . splitOn ": "

determineWinners :: [([Int], [Int])] -> [Int]
determineWinners = map ((2 ^) . (1 `subtract`)) . filter (/= 0) . map (length . uncurry intersect)

processCards :: Map Int ([Int], [Int]) -> [Int]
processCards m = go $ M.keys m
  where
    cardsWon = M.mapWithKey (\k v -> [k + 1 .. k + v]) $ M.filter (/= 0) $ M.map (length . uncurry intersect) m
    go :: [Int] -> [Int]
    go [] = []
    go xs = xs <> go (concat $ mapMaybe (`M.lookup` cardsWon) xs)

main' :: IO ()
main' = do
  contents <- lines <$> readFile "../inputs/2023/Day4/input.txt"
  let cards = M.fromList $ map parseCard contents
  -- part 1
  let winners = determineWinners $ M.elems cards
  print $ sum winners
  -- part 2
  let totalCards = processCards cards
  print $ length totalCards
  return ()

File: typescript/2023/day4.ts
import { drop, intersection, isEmpty, isNotNil, range, sum } from 'ramda';

const content = (await Bun.file('../../inputs/2023/Day4/input.txt').text()).split('\n').filter(s => !isEmpty(s));

const parseNumbers = (str: string) =>
  str
    .split(' ')
    .filter(x => !isEmpty(x))
    .map(x => Number.parseInt(x, 10));

const parseCard = (cardStr: string) => {
  const [cardName, rest] = cardStr.split(': ');
  const cardNum = Number.parseInt(drop(5, cardName), 10);
  const [winners, holding] = rest.split(' | ').map(parseNumbers);

  return { cardNum, holding, winners };
};

type Card = ReturnType<typeof parseCard>;

const parsed: Card[] = content.map(parseCard);

const determineWinners = (cards: Card[]) =>
  cards
    .map(card => intersection(card.winners, card.holding).length)
    .filter(x => x !== 0)
    .map(x => 2 ** (x - 1));

// part 1
console.log(sum(determineWinners(parsed)));

const processCards = (cards: Card[]): number[] => {
  const cardsWon = Object.fromEntries(
    cards
      .map(card => [card.cardNum, intersection(card.winners, card.holding).length] as [number, number])
      .filter(([, l]) => l !== 0)
      .map(([k, v]) => [k, range(k + 1, k + v + 1)] as [number, number[]]),
  );

  let acc: number[] = cards.map(x => x.cardNum);
  let xs: number[] = cards.map(x => x.cardNum);

  while (xs.length) {
    xs = xs
      .map(x => cardsWon[x])
      .filter(isNotNil)
      .flat();

    acc = acc.concat(xs);
  }

  return acc;
};

// part 2
console.log(processCards(parsed).length);
